### ISSUE TITLE
Replicating multiple pools

### ISSUE THREAD

**Issue by mlen** [alunduil/zfs-replicate#92]
This is a really awesome project and it was super easy to set it up for my laptop. This isn't as obvious for hosts that import multiple pools – it seems I can only replicate one of them if I want to use the config DSL in nixos. Did you consider adding support for multiple pools?

This can technically happen either here or in nixpkgs – not sure which approach is cleaner.

**Comment by alunduil** [comment]
Thanks!  I did struggle with whether or not to use submodules in the NixOS configuration module when I first wrote it but since I didn't have multiple pools I didn't know what the best way to express it might be.  The last little irksome detail was the names for the systemd units and whether or not to use templating for those (anything I looked at seemed super unreadable and so I punted).

I still think submodules makes the most sense but would need some help structuring that change.  That keeps this software simple and light with just shuffling data from one place to another.

What do you think?  If you like the idea of submodules and are open to experimenting with it a bit, we can pick this up in a pull request on nixos/nixpkgs instead.  Let me know what you think.

**Comment by mlen** [comment]
What would you say about structuring the configuration like this?
```
services.zfs.autoReplication = {
  enable = true;
  filesystems = {
    "pool/local/fs" = {
      host = "some.hostname";
      username = "root";
      identityFilePath = "/etc/nixos/secret/zfs-replication";
      remoteFilesystem = "backup";
    };
    "tank/some/other/fs" = {
      host = "some.hostname";
      username = "root";
      identityFilePath = "/etc/nixos/secret/zfs-replication";
      remoteFilesystem = "backup";
    };
  };
};
```

I think it would be the best to generate multiple service files -- this allows to run the tasks in parallel and also it's easier to see which part failed. Maybe it would be possible to group these in a systemd target unit?

**Comment by alunduil** [comment]
I like that layout for the NixOS configuration and it's pretty much what I was thinking last time.  One thing I can think of that we can punt on again is what if I want to replicate one filesystem to two destinations?

We definitely want multiple systemd units (one per filesystem to replicate).  Due to the way I hook them into the existing zfs-snapshot units I don't think a target makes sense, but I could be convinced that the target should be pulled in by those other units instead as long as the target will re-run all of its units when triggered (I can't remember if targets will do that or not).

The other thing I'm worried about with the units is having a readable structure in the names.  Do we name the units zfs-replicate-fs@destination or zfs-replicate-fs-destination or something even weirder and possibly unreadable?  A possibly nice answer is to use the hash and skip the names so they at least all have the same length and stay somewhat legible when printed out.  What do you think about this problem?

**Comment by alunduil** [comment]
@mlen thanks for the report and sorry for the silence, this was closed by the stale bot rather than by anyone deciding anything. The CLI half of what you asked for now has concrete successors: #395 for replicating multiple local to remote pairs in one invocation via a repeatable `--pair LOCAL:REMOTE`, and #396 for a TOML config so those pairs live in a file instead of a long command line. The NixOS module layout we sketched out here is really a nixpkgs question rather than something this repo can carry, so I've left that part out of both. Marking this a duplicate of #395.
